src/contract.ts

```diff
@@ -0,0 +1,78 @@
+import { provider } from './config';
+
+import * as BASE_REGISTRAR_BYTECODE from './assets/ENSRegistrar_bc.json';
+import * as NAME_WRAPPER_BYTECODE from './assets/NameWrapper_bc.json';
+import { ethers } from 'ethers';
+import { Version } from './metadata';
+
+
+export interface BaseError {}
+export class BaseError extends Error {
+  __proto__: Error;
+  constructor(message?: string) {
+    const trueProto = new.target.prototype;
+    super(message);
+
+    this.__proto__ = trueProto;
+  }
+}
+
+export interface ContractMismatchError {}
+export class ContractMismatchError extends BaseError {}
+
+export interface ContractNotFoundError {}
+export class ContractNotFoundError extends BaseError {}
+
+export interface OwnerNotFoundError {}
+export class OwnerNotFoundError extends BaseError {}
+
+
+export async function checkContract(
+  contractAddress: string,
+  tokenId: string
+): Promise<Version> {
+  let nftOwner;
+  const contract_code = await provider.getCode(contractAddress);
```

Comment:
Do you need to call `.getCode` for every call? That probably increases our infura cost unless you add caching. 

v1 contract address can be fetched when the metadata service starts by calling `ens.owner('eth')`.
v2 is a bit tricky but you can potentially do something like `NameWrapper.supportsInterface(interfaceid)`.

Getting the interfaceid of NameWrapper is a bit tricky as it doesn't expose the interfaceid directly. I manually copy&pasted the code into remix, add the following code to get the value.

```
    function getInterfaceId() external pure returns (bytes4){
        return type(INameWrapper).interfaceId;
    }
```

The interfaceid of the latest code is `0x1aa28a1e`.
The interfaceid of `https://rinkeby.etherscan.io/address/0x9029c1574f91696026358d4edB0De773d0E04aeD` (the one I deployed a while back) is `0xa0ab9076`.

Reply:
Possibly the easiest is just to hardcode the wrapper address at config params  (we need to hardcode wrapper address on subgraph config as well) and do something like this. You can also hardcode .eth registrar address, but that can easily be fetched dynamically by calling `ens.owner('eth')` once

```js
try{
  nftOwner = await contract.ownerOf(tokenId);
}catch(e){
 // tokenId does not exist
}
if(contract.address === NAME_WRAPPER_ADDRESS){
  return Version.v2
}else if (contract.address === ETH_REGISTRAR_ADDRESS){
  if(nftOwner === ETH_REGISTRAR_ADDRESS){
    return Version.v1w
  }else{
    return Version.v1
  }
}else{
   throw('this contract is not supported')
}

```

Reply:
Sure, I will take a look for this option. The reason I wanted to check for bytecode, was trying to find a general verification process for whichever address the contract is residing unless someone modifies it. But of course we can go with hardcode option as well for an ease.